feat: store min, max, null count, and true count in column metadata

danking opened the pull request.

lwwmanning committed:
wip on stats

lwwmanning committed:
write stats

lwwmanning committed:
cosmetic change

lwwmanning committed:
trying to do this with rust types is awful

lwwmanning committed:
wip

lwwmanning committed:
cleanup

danking committed:
feat: store min, max, null count, and true count in column metadata

danking commented:
fyi @lwwmanning I rebased your PR on develop.

danking commented:
Alright, this does not break the current read/write tests (which might only be in Python doc tests?). We need to actually read and use the data to really test it though.

lwwmanning commented on vortex-serde/src/layouts/write/metadata_accumulators.rs:
```diff
@@ -0,0 +1,263 @@
+    minima: Vec<Option<T>>,
+    maxima: Vec<Option<T>>,
+    to_array: fn(Vec<Option<T>>) -> Array,
+    basic_metadata: Box<BasicAccumulator>,
```
This doesn't need to be a box?

danking replied:
Done.

lwwmanning commented on vortex-serde/src/layouts/write/metadata_accumulators.rs:
```diff
@@ -0,0 +1,263 @@
+        ))
+    }
+
+    fn into_layouts_and_metadata_parts(
```
This is never called, right? Can make `into_layouts_and_metadata_parts` a fn on `BasicAccumulator` instead of the trait

danking replied:
Done.

danking committed:
address comments

robert3005 commented on vortex-serde/src/layouts/write/metadata_accumulators.rs:
```diff
@@ -0,0 +1,256 @@
+            );
+        }
+
+        let mut names: Vec<String> = vec!["row_offset".into()];
```
If you make this `Vec<Arc<str>>` you will avoid the annoying conversion function

danking replied:
Done.

danking committed:
address rob comments

danking committed:
clippy

robert3005 commented:
@danking i will do a proper review in ~1h

robert3005 commented on vortex-serde/src/layouts/write/writer.rs:
```diff
@@ -258,6 +226,5 @@ mod tests {
             buffer[buffer_begin..buffer_end].len(),
             FOOTER_POSTSCRIPT_SIZE
         );
-        assert_eq!(buffer[buffer_begin..buffer_end].len(), 32);
```
This asserts that we don't accidentally change `FOOTER_POSTSCRIPT_SIZE`

danking replied:
Done.

robert3005 commented on vortex-serde/src/layouts/write/writer.rs:
```diff
@@ -83,74 +82,58 @@ impl<W: VortexWrite> LayoutWriter<W> {
     async fn write_column_chunks<S>(&mut self, mut stream: S, column_idx: usize) -> VortexResult<()>
     where
-        S: Stream<Item = VortexResult<Array>> + Unpin,
+        S: Stream<Item = VortexResult<Array>> + Unpin + ArrayStream,
```
I guess ArrayStream implies Stream? Can get rid of the earlier bound

danking replied:
Done.

robert3005 commented on vortex-serde/src/layouts/write/metadata_accumulators.rs:
```diff
@@ -0,0 +1,255 @@
+}
+
+impl<T> ExtremaAccumulator<T> {
+    fn new(size_hint: usize, to_array: fn(Vec<Option<T>>) -> Array) -> Self {
```
It's weird to pass this function here directly. All you want is `impl Into<Array> for ExtremaAccumulator<...>`

robert3005 commented on vortex-serde/src/layouts/write/metadata_accumulators.rs:
```diff
@@ -0,0 +1,255 @@
+
+    fn push_batch_byte_offsets(&mut self, batch_byte_offsets: Vec<u64>);
+
+    fn into_layouts_and_metadata(self: Box<Self>) -> VortexResult<(VecDeque<Layout>, StructArray)>;
```
If you split this method into separate trait you can avoid passing the conversion function and instead specialize as necessary. 

```
pub trait IntoArrayMetadata {
    fn into_array_metadata(self: Box<Self>) -> VortexResult<(VecDeque<Layout>, StructArray)>;
}
```

robert3005 reviewed (changes requested):
Explicitly passing the conversion function seems like a hack. I think we would also like to specialize BoolMetadataAccumulator to hold true counts since they don't appear otherwise

danking committed:
maybe address comments?

danking committed:
fix up the writer too

danking committed:
final cleanup of writer and accumulators

danking commented:
Erm. Ok. I might have gone rogue here so push back, if so. 

I implemented `From<Vec<Option<T>>>` for the standard types that I need. I define three accumulators: BoolAccumulator, StandardAccumulator (has extrema), and BasicAccumulator (just row offsets and null counts).

I also switched the writer.rs to use a new `ColumnWriter` which encapsulates all the per-column work.

danking commented on vortex-serde/src/layouts/write/metadata_accumulators.rs:
```diff
@@ -0,0 +1,243 @@
+
+pub fn new_metadata_accumulator(dtype: &DType) -> Box<dyn MetadataAccumulator> {
+    match dtype {
+        DType::Null => Box::new(BasicAccumulator::new()),
```
It occurs to me that we could probably also specialize all these to have/not-have null counts based on the nullability.

robert3005 replied:
I think we can defer this one for later. I’m not sure the null ability matters for performance here

robert3005 commented on vortex-serde/src/layouts/write/writer.rs:
```diff
@@ -200,19 +160,76 @@ async fn write_fb_raw<W: VortexWrite, F: WriteFlatBuffer>(mut writer: W, fb: F)
+        Ok(())
+    }
+
+    async fn write_into<W: VortexWrite>(
```
This should be something like `write_metadata`?

lwwmanning committed:
clippy + added some metadata schema tests

lwwmanning committed:
Merge remote-tracking branch 'origin/develop' into dk/write-stats

lwwmanning committed:
better stats coverage
